### ISSUE TITLE
StackOverflowException when using SDK_UnityController

### ISSUE THREAD

**Issue by MattOstgard** [thestonefox/VRTK#1764]
### Environment

 * Source of VRTK: GitHub
 * Version of VRTK 3.3.0-alpha, 72c9f432c84abea285c5009f637951a07181ca6f
 * Version of Unity 2017.4.0f1
 * Hardware used Oculus Utilities v1.24.0 (2018-03-12)
 * SDK used: Unity SDK for Standalone:Oculus

### Steps to reproduce

  1. Set SDK Setup on the UnityXR gameobject to `Unity (Standalone:Oculus)`
  2. Start the game and switch to UnityXR SDK after which the exception should be thrown.

### Expected behavior

Not to get stuck in a loop.

### Current behavior

This exception is thrown:
<details><summary>StackOverflowException ...</summary>

```
VRTK.SDK_BaseHeadset.ScrapeHeadsetType () (at Assets/VRTK/Source/SDK/Base/SDK_BaseHeadset.cs:150)
VRTK.SDK_UnityHeadset.GetHeadsetType () (at Assets/VRTK/Source/SDK/Unity/SDK_UnityHeadset.cs:69)
VRTK.VRTK_SDK_Bridge.GetHeadsetType () (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:209)
VRTK.VRTK_DeviceFinder.GetHeadsetTypeAsString () (at Assets/VRTK/Source/Scripts/Utilities/VRTK_DeviceFinder.cs:477)
VRTK.VRTK_DeviceFinder.GetHeadsetType () (at Assets/VRTK/Source/Scripts/Utilities/VRTK_DeviceFinder.cs:486)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:721)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControllerButtons (ControllerHand hand) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:742)
VRTK.SDK_UnityController.SetTrackedControllerCaches (Boolean forceRefresh) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:696)
VRTK.SDK_UnityController.GetTrackedObject (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:810)
VRTK.SDK_UnityController.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/Unity/SDK_UnityController.cs:190)
VRTK.VRTK_SDK_Bridge.GetControllerIndex (UnityEngine.GameObject controller) (at Assets/VRTK/Source/SDK/VRTK_SDK_Bridge.cs:42)
VRTK.VRTK_ControllerReference.GetControllerReference (ControllerHand controllerHand) (at Assets/VRTK/Source/Scripts/Internal/VRTK_ControllerReference.cs:47)
VRTK.SDK_UnityController.SetControll<message truncated>
```
</details>

**Comment by thestonefox** [comment]
I've tried this in Unity 2017.3.0p1 and can't recreate the issue. I'll try in 2017.4 as well to see if there was any change

**Comment by thestonefox** [comment]
I've tried in 2017.4 and still can't recreate this

**Comment by MattOstgard** [comment]
Very strange. I thought maybe there was something important I left out of my repro steps so went through them again and was able to reproduce even easier some how. Here is what I did:

1) Downloaded the alpha branch from GitHub via `Clone & Download` > `Download .ZIP`
2) Extracted `VRTK-release-3.3.0-alpha.zip`
3) Opened the project in Unity 2017.4.0f1 via Unity Hub with windows selected as the build target
4) Confirmed "Opening project in non-matching editor installation" dialog
5) Opened the scene `VRTK/Examples/VRTK_SDKManager_Constructor.unity`
6) Pressed the play button

SteamVR starts up, the sdk switcher auto switches to UnityXR, and then in the Unity console the stack overflow exception is produced. This is without changing anything to VRTK at all. As a hail mary I also tried installing the SteamVR Plugin off of the Unity Asset Store, but of course there was no difference.

I'm not sure what is going on here. I'll try to poke around and figure out why it is breaking for me, but not sure when that will be.

**Comment by tom-fv** [comment]
We've had this issue as well, we're using 2017.4.1f1, with only the Unity SDK installed, no others installed, using a HTC Vive, seems to be an infinite loop of finding the controller reference on the last line of SetControllerButtons of the SDK_UnityController.cs, so it might be to do when you don't have one hand of the headset enabled perhaps?

**Comment by jamre** [comment]
I've been getting this too. (2017.4.2f2), 2364628
In my own project, I only have the UnityXR SDK.
In the example project, I get the exception if I move the UnityXR SDK to the top:
![image](https://user-images.githubusercontent.com/1468331/39782307-67698c34-52d7-11e8-9e30-96c80bb6bcda.png)

I found a few issues that are causing it:
**SetTrackedControllerCaches** calls **SetControllerButtons**. If the joystick is not found, it calls **VRTK_ControllerReference.GetControllerReference** (twice), which calls **SetTrackedControllerCaches** again each time, putting it into infinite recursion.

The reason the joysticks are not found is the names do not match. I changed my valid hands lists to include the names that Unity is reporting:

```cs
protected List<string> validRightHands = new List<string>()
{
    "OpenVR Controller - Right",
    "OpenVR Controller(Vive. Controller MV) - Right",
    "Oculus Touch - Right",
    "Oculus Remote"
};
protected List<string> validLeftHands = new List<string>()
{
    "OpenVR Controller - Left",
    "OpenVR Controller(Vive. Controller MV) - Left",
    "Oculus Touch - Left"
};
```

This change, along with a flag to not allow reentrancy into the **SetTrackedControllerCaches** method and moving the **SetControllerButtons** calls into the null check seem to have fixed the issue for me.

The reason I had to move the SetControllerButtons calls into the null check is that 
```cs
VRTK_ControllerReference.GetControllerReference(hand).actual.gameObject.activeInHierarchy
```
would throw a null reference exception otherwise.
